docs/error_handling.md

```diff
@@ -0,0 +1,200 @@
+but if you need to take the address the Raising proc, please add the
+`addressable` pragma that will force the wrapper to be a regular proc.
+
+Finally, `failing` is another pragma provided for convenience which
+is equivalent to `{.errors: (CatchableError).}`.
+
+## The `raising` annotation
+
+The `raising` annotation is the simplest form of error tracking
+similar to the `try` annotation made popular by the [Midori error model](http://joeduffyblog.com/2016/02/07/the-error-model/),
+which is also [proposed for addition in C++](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf) and [already available in Zig](https://ziglang.org/documentation/master/#try).
+
+It merely marks the locations in the code where errors might be raised
+and strips away the `Raising` type to disarm the compiler checks:
+
+```nim
+proc attachValidator(node: BeaconNode, keyFile: string) {.failing.} =
+  node.addValidator(raising ValidatorPrivKey.init(raising readFile(keyFile))
+```
+
+When applied to a `Result` or an `Option`, `raising` will use the
+`tryGet` API to attempt to obtain the computed value.
+
+## The capital `Try` API
+
+The capital `Try` API is similar to a regular `try` expression
+or a `try` statement. The only difference is that you must provide
+exception handlers for all possible recoverable errors. If you fail
+to do so, the compiler will point out the line in the `try` block
+where an unhandled exception might be raised:
+
+```nim
+proc replaceValues(values: var openarray[string],
+                   replacements: Table[MyEnum, string]) =
+  Try:
+    for v in mitems(values):
+      let
+        enumValue = parseEnum v
+        replacement = replacements[enumValue] # Error here
+      v = replacement
+  except ValueError:
+    echo "Invalid enum value"
+```
+
+The above example will fail to compile with an error indicating
+that `replacements[enumValue]` may fail with an unhandled `KeyError`.
+
+## The `either` expression
+
+The `either` expression can be used with APIs based on `Option[T]`,
+`Result[T, E]` or the `errors` pragma when it's appropriate to
+discriminate only between successful execution and any type of
+failure. Regardless of the error handling scheme being used,
+`either` is used like this:
+
+```nim
+let x = either(foo(), fallbackValue)
+```
+
+On success, `either` returns the successfully computed value
+and the failure side of the expression won't be evaluated at all.
+
+Besides providing a substitute value, the failure side of the
+expression may also feature a `noReturn` statement such as
+`return`, `raise`, `quit` as long at it's used with the following
+syntax:
+
+```nim
+let x = either foo():
+               return
+```
+
+Within the failure path, you can also use the `error` keyword to
+refer to the raised exception or the `error` value of the failed
+`Result`.
+
+## The `check` expression
+
+The `check` macro provides a general mechanism for handling
```

Comment:
so we don't call it `match` because... it's not general enough?

Reply:
`check` is about unpacking the successful result while handling the errors. I think that's different enough from general pattern matching to deserve its own name.